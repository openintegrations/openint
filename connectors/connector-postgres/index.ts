// Should this actually be called `connector-export` or `connector-standard`? Similar to `connector-spreadsheet`?
// Previously this used to be called `connector-alka`, but given alka is no longer relevant, how do we want to define this?

import postgresDef from './def'
import postgresServer from './server'

// codegen:start {preset: barrel, include: "./{*.{ts,tsx},*/index.{ts,tsx}}", exclude: "./**/*.{d,spec,test,fixture,gen,node}.{ts,tsx}"}
export * from './def'
export * from './makePostgresClient'
export * from './server'
// codegen:end

export const postgresProvider = {
  ...postgresDef,
  ...postgresServer,
}
