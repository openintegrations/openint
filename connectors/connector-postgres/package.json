{
  "name": "@openint/connector-postgres",
  "version": "0.0.0",
  "private": true,
  "module": "./index.ts",
  "dependencies": {
    "@openint/cdk": "workspace:*",
    "@openint/util": "workspace:*",
    "@slonik/migrator": "0.11.3",
    "handlebars": "4.7.7",
    "pg": "8.8.0",
    "pg-native": "3.0.1",
    "slonik": "30.3.1",
    "slonik-interceptor-preset": "1.2.10",
    "zod": "3.22.4"
  },
  "devDependencies": {
    "@types/pg": "8.6.5"
  }
}
