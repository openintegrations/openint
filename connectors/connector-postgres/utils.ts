import {sql} from 'drizzle-orm'
import type {PgColumn, PgDatabase} from 'drizzle-orm/pg-core'
import {pgTable, uniqueIndex} from 'drizzle-orm/pg-core'
import type {RecordMessageBody} from './def'

export function isValidDateString(str: string) {
  const date = new Date(str)
  const lenToCheck = Math.min(str.length, '2021-01-01T00:00:00'.length)
  return (
    !Number.isNaN(date.getTime()) &&
    date.toISOString().slice(0, lenToCheck) === str.slice(0, lenToCheck)
  )
}

/**
 * We probably want something based off of the Catalog protocol, by inferring catalog
 * from data object and then generating a migration from the catalog message as a two step,
 * unified process.
 * For now we just do something lightweig thought
 *
 * Also consider deduping with inferTableForUpsert
 */
export function inferTableFromMessage(event: RecordMessageBody) {
  // TODO Implement namespace and upsert metadata support
  return pgTable(
    event.stream,
    (t) => {
      function inferCol(v: unknown) {
        if (typeof v === 'string') {
          return isValidDateString(v) ? t.timestamp({mode: 'string'}) : t.text()
        }
        if (typeof v === 'number') {
          return Math.floor(v) === v ? t.integer() : t.doublePrecision()
        }
        if (typeof v === 'boolean') {
          return t.boolean()
        }
        if (v instanceof Date) {
          return t.timestamp()
        }
        return t.jsonb()
      }
      return Object.fromEntries(
        Object.entries(event.data).map(([k, v]) => [k, inferCol(v)]),
      )
    },
    (table) =>
      // TODO: Check against existing primary key and unique indexes
      // to not create if already there?
      // && Object.values(table).filter(c => c.i)
      event.upsert?.key_columns
        ? [
            uniqueIndex(`${event.stream}_upsert_keys`).on(
              ...(event.upsert.key_columns.map(
                (col) => table[col] as PgColumn,
              ) as [PgColumn]),
            ),
          ]
        : [],
  )
}

export async function runMigrationForStandardTable(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  db: PgDatabase<any, any>,
  tableName: string,
) {
  // TODO: turn this into a drizzle table
  // Where do we want to put data? Not always public...
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return await db.execute(sql`
    CREATE TABLE IF NOT EXISTS ${sql.identifier(tableName)} (
      source_id VARCHAR NOT NULL,
      id VARCHAR NOT NULL,
      customer_id VARCHAR,
      created_at timestamp with time zone DEFAULT now() NOT NULL,
      updated_at timestamp with time zone DEFAULT now() NOT NULL,
      connector_name VARCHAR GENERATED ALWAYS AS (split_part((source_id)::text, '_'::text, 2)) STORED NOT NULL,
      CONSTRAINT ${sql.identifier(
        `pk_${tableName}`,
      )} PRIMARY KEY ("source_id", "id"),
      unified jsonb,
      raw jsonb DEFAULT '{}'::jsonb NOT NULL
    );
  `)
  // NOTE: Should we add org_id?
  // NOTE: Rename `unified` to `unified` and `raw` to `raw` or `remote` or `original`
  // NOTE: add prefix check would be nice
  // for (const col of [
  //   'id',
  //   'source_id',
  //   'connector_name',
  //   'created_at',
  //   'updated_at',
  //   'customer_id',
  // ]) {
  //   await pool.query(sql`
  //     CREATE INDEX IF NOT EXISTS ${sql.identifier([
  //       `${tableName}_${col}`,
  //     ])} ON ${table} (${sql.identifier([col])});
  //   `)
  // }
}
