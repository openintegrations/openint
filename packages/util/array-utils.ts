import type {MaybeArray, NonEmptyArray} from './type-utils'

export {inPlaceSort, sort} from 'fast-sort'

export function nonEmpty<T>(arr: T[]) {
  if (arr.length === 0) {
    throw new Error('Expected non-empty array')
  }
  return arr as NonEmptyArray<T>
}

/**
 * https://stackoverflow.com/a/16436975/692499
 */
export function arrayEquals<T>(
  a: T[],
  b: T[],
  isEqual: (a: T | undefined, b: T | undefined) => boolean = Object.is,
) {
  if (a === b) {
    return true
  }
  if (a.length !== b.length) {
    return false
  }
  for (let i = 0; i < a.length; ++i) {
    if (!isEqual(a[i], b[i])) {
      return false
    }
  }
  return true
}

export function fromMaybeArray<T>(maybeArray: MaybeArray<T>) {
  return Array.isArray(maybeArray) ? maybeArray : [maybeArray]
}

/**
 * Removes all falsy values (false, null, 0, "", undefined, and NaN) from an array
 */
export function compact<T>(
  array: Array<T | false | null | undefined | 0 | '' | 0n>,
) {
  return array.filter(Boolean) as T[]
}
