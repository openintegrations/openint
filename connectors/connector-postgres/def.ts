import type {
  ConnectorDef,
  ConnectorSchemas,
  EntityPayloadWithRaw,
} from '@openint/cdk'
import {connHelpers} from '@openint/cdk'
import {z, zCast} from '@openint/util'

export const zPgConfig = z.object({
  databaseUrl: z.string(),
  migrationsPath: z.string().optional(),
  migrationTableName: z.string().optional(),
  transformFieldNames: z.boolean().optional(),
  migrateTables: z.boolean().optional(),
})

export const postgresSchemas = {
  name: z.literal('postgres'),
  // TODO: Should postgres use integration config or resourceSettings?
  // if it's resourceSettings then it doesn't make as much sense to configure
  // in the list of integrations...
  // How do we create default resources for integrations that are basically single resource?
  resourceSettings: zPgConfig.pick({databaseUrl: true, migrateTables: true}).extend({
    // gotta make sourceQueries a Textarea

    sourceQueries: z
      .object({
        invoice: z
          .string()
          .nullish()
          .describe('Should order by lastModifiedAt and id descending'),
      })
      // .nullish() does not translate well to jsonSchema
      // @see https://share.cleanshot.com/w0KVx1Y2
      .optional(),
  }),
  destinationInputEntity: z.object({
    id: z.string(),
    entityName: z.string(),
    // TODO: Fix the support here. We hare hacking postgres to be able
    // support both unified +unified inputs and raw only inputs
    // Basically this should work with or without a link... And it's hard to abstract for now
    entity: z.object({
      // For now... in future we shall support arbitrary columns later
      raw: z.any(),
      unified: z.any(),
    }),
  }),
  sourceOutputEntity: zCast<EntityPayloadWithRaw>(),
  sourceState: z
    .object({
      invoice: z
        .object({
          lastModifiedAt: z.string().optional(),
          lastRowId: z.string().optional(),
        })
        .optional(),
    })
    .optional(),
} satisfies ConnectorSchemas

export const postgresHelpers = connHelpers(postgresSchemas)

export const postgresDef = {
  name: 'postgres',
  metadata: {
    verticals: ['database'],
    logoUrl: '/_assets/logo-postgres.png',
    stage: 'ga',
  },

  schemas: postgresSchemas,
  standardMappers: {
    resource: (_settings) => ({
      displayName: 'Postgres',
      status: 'healthy',
    }),
  },
} satisfies ConnectorDef<typeof postgresSchemas>

export default postgresDef
