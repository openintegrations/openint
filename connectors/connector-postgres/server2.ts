/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import {generateDrizzleJson, generateMigration} from 'drizzle-kit/api'
import type {PgColumn, PgDatabase, PgTable} from 'drizzle-orm/pg-core'
import {pgTable, uniqueIndex} from 'drizzle-orm/pg-core'
import jsonStableStringify from 'json-stable-stringify'
import * as R from 'remeda'
import {z} from 'zod'
import type {ConnectorServer} from '@openint/cdk'
import {handlersLink} from '@openint/cdk'
import {dbUpsert} from '@openint/db'
import {rxjs} from '@openint/util'
import type {postgresSchemas} from './def'
import {isValidDateString} from './utils'

export const postgresServer = {
  destinationSync: ({endUser, source, settings: {databaseUrl}}) => {
    return handlersLink({
      data: async (op) => {
        return rxjs.of(op)
      },
      commit: async (op) => {
        return op
      },
    })
  },
} satisfies ConnectorServer<typeof postgresSchemas>

export default postgresServer

/** Airbyte record message */
export const zRecordMessage = z.object({
  stream: z.string(),
  // Should we support non-record data?
  data: z.record(z.unknown()),
  namespace: z.string().optional(),
  upsert: z
    .object({
      insert_only_columns: z.array(z.string()).optional(),
      key_columns: z.array(z.string()).optional(),
      must_match_columns: z.array(z.string()).optional(),
      no_diff_columns: z.array(z.string()).optional(),
      shallow_merge_jsonb_columns: z.array(z.string()).optional(),
    })
    .optional(),
})

export type RecordMessage = z.infer<typeof zRecordMessage>

/**
 * We probably want something based off of the Catalog protocol, by inferring catalog
 * from data object and then generating a migration from the catalog message as a two step,
 * unified process.
 * For now we just do something lightweig thought
 */
export function inferTable(event: RecordMessage) {
  // TODO Implement namespace and upsert metadata support
  return pgTable(
    event.stream,
    (t) => {
      function inferCol(v: unknown) {
        if (typeof v === 'string') {
          return isValidDateString(v) ? t.timestamp() : t.text()
        }
        if (typeof v === 'number') {
          return Math.floor(v) === v ? t.integer() : t.doublePrecision()
        }
        if (typeof v === 'boolean') {
          return t.boolean()
        }
        if (v instanceof Date) {
          return t.timestamp()
        }
        return t.jsonb()
      }
      return Object.fromEntries(
        Object.entries(event.data).map(([k, v]) => [k, inferCol(v)]),
      )
    },
    (table) =>
      // TODO: Check against existing primary key and unique indexes
      // to not create if already there?
      // && Object.values(table).filter(c => c.i)
      event.upsert?.key_columns
        ? [
            uniqueIndex(`${event.stream}_upsert_keys`).on(
              ...(event.upsert.key_columns.map(
                (col) => table[col] as PgColumn,
              ) as [PgColumn]),
            ),
          ]
        : [],
  )
}

export function getMigrationsForTable(table: PgTable) {
  return generateMigration(
    generateDrizzleJson({}),
    generateDrizzleJson({table}),
  )
}

export async function upsertFromRecordMessages(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  db: PgDatabase<any, any>,
  messages: RecordMessage[],
) {
  const messagesByConfig = R.groupBy(messages, (m) =>
    jsonStableStringify(R.pick(m, ['stream', 'namespace', 'upsert'])),
  )
  for (const msgs of Object.values(messagesByConfig)) {
    const table = inferTable(msgs[0])

    // This should be cached within a single destinationSync run and therefore
    // the interface of upsertFromRecordMessages may change...
    const migrations = await getMigrationsForTable(table)
    for (const migration of migrations) {
      await db.execute(migration)
    }
    // This doesn't work because upsert needs reference to table column to know how to
    // serialize value, For example string into jsonb is not the same as string into a text
    // column. So it's actually somewhat important to get reference to the table schema
    // rather than needing to infer it each time.
    // const table2 = pgTable(
    //   message.stream,
    //   (t) =>
    //     Object.fromEntries(
    //       Object.keys(message.data).map((k) => [k, t.jsonb()]),
    //     ) as Record<string, any>,
    // )
    if (msgs[0].upsert) {
      await dbUpsert(
        db,
        table,
        msgs.map((m) => m.data),
        {
          insertOnlyColumns: msgs[0].upsert?.insert_only_columns,
          keyColumns: msgs[0].upsert?.key_columns,
          mustMatchColumns: msgs[0].upsert?.must_match_columns,
          noDiffColumns: msgs[0].upsert?.no_diff_columns,
          shallowMergeJsonbColumns: msgs[0].upsert?.shallow_merge_jsonb_columns,
        },
      )
    } else {
      await db.insert(table).values(msgs.map((m) => m.data))
    }
  }
  return messagesByConfig
}
