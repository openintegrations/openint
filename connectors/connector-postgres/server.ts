/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import type {DatabasePool} from 'slonik'
import {sql} from 'slonik'
import type {ConnectorServer} from '@openint/cdk'
import {handlersLink} from '@openint/cdk'
import {R, Rx, rxjs, snakeCase} from '@openint/util'
import type {postgresSchemas} from './def'
import {postgresHelpers} from './def'
import {makePostgresClient, upsertByIdQuery} from './makePostgresClient'

// TODO: remove when we introduce dynamic column names
const agTableMappings = [
  {from: 'integration_ats_job', to: 'IntegrationATSJob'},
  {from: 'integration_ats_candidate', to: 'IntegrationATSCandidate'},
  {from: 'integration_ats_job_opening', to: 'IntegrationATSJobOpening'},
  {from: 'integration_ats_offer', to: 'IntegrationATSOffer'},
  {from: 'integration_connection', to: 'IntegrationConnection'},
  {from: 'integration_ats_opening', to: 'IntegrationATSOpening'},
]

async function setupTable({
  pool,
  schema: _schema,
  tableName: _tableName,
}: {
  pool: DatabasePool
  schema?: string
  tableName: string
}) {
  const schema = snakeCase(_schema)
  const tableName = snakeCase(_tableName)
  // TODO: remove when we introduce dynamic column names
  const mappedTableName =
    agTableMappings.find((mapping) => mapping.from === tableName)?.to ||
    tableName
  const table = sql.identifier(
    schema ? [schema, mappedTableName] : [mappedTableName],
  )

  // Note: Ideally we would create these for AG but we're not doing so as slonik does not support raw queries
  // and also these tables are precreated in staging
  // import { raw } from 'slonik-sql-tag-raw';
  // const extraPerEntityColumns = {
  //   'IntegrationATSJob': ['external_job_id VARCHAR'],
  //   'IntegrationATSCandidate': ['opening_external_id VARCHAR', 'candidate_name VARCHAR'],
  //   'IntegrationATSJobOpening': ['opening_external_id VARCHAR', 'job_id VARCHAR'],
  //   'IntegrationATSOffer': ['opening_external_id VARCHAR', 'candidate_name VARCHAR'],
  // }
  // const extraColumns = extraPerEntityColumns[mappedTableName as keyof typeof extraPerEntityColumns]?.join(',\n      ') ?? ''
  // this does not work... 
  // const extraColumnsSql = extraColumns.length > 0 ? sql`, ${raw(extraColumns)}` : sql``;

  await pool.query(sql`
    CREATE TABLE IF NOT EXISTS ${table} (
      "connectionId" VARCHAR NOT NULL,
      id VARCHAR NOT NULL,
      "clientId" VARCHAR,
      "createdAt" timestamp with time zone DEFAULT now() NOT NULL,
      "updatedAt" timestamp with time zone DEFAULT now() NOT NULL,
      "connectorName" VARCHAR GENERATED ALWAYS AS (split_part(("connectionId")::text, '_'::text, 2)) STORED NOT NULL,
      CONSTRAINT ${sql.identifier([`pk_${mappedTableName}`])} PRIMARY KEY (
        "id"),
      unified jsonb,
      raw jsonb DEFAULT '{}'::jsonb NOT NULL,
      "isOpenInt" boolean
    );
  `)
  // NOTE: Should we add org_id?
  // NOTE: Rename `unified` to `unified` and `raw` to `raw` or `remote` or `original`
  // NOTE: add prefix check would be nice
  for (const col of [
    'id',
    'connectionId',
    // 'connectorName', // TODO: remove when we introduce dynamic column names
    'createdAt',
    'updatedAt',
    'clientId',
  ]) {
    await pool.query(sql`
      CREATE INDEX IF NOT EXISTS ${sql.identifier([
        `${mappedTableName}_${col}`,
      ])} ON ${table} (${sql.identifier([col])});
    `)
  }
}

export const postgresServer = {
  // TODO:
  // 1) Implement pagination
  // 2) Impelemnt incremental Sync
  // 3) Improve type safety
  // 4) Implement parallel runs
  sourceSync: ({
    endUser,
    settings: {databaseUrl, sourceQueries},
    state = {},
  }) => {
    const {getPool} = makePostgresClient({databaseUrl})
    // TODO: Never let slonik transform the field names...
    const rawClient = makePostgresClient({
      databaseUrl,
      transformFieldNames: false,
    })

    async function* iterateEntities() {
      const handlebars = await import('handlebars')

      const pool = await getPool()
      // Where do we want to put data? Not always public...
      await setupTable({pool, tableName: 'account'})
      await setupTable({pool, tableName: 'transaction'})

      for (const entityName of ['account', 'transaction'] as const) {
        const res = await pool.query<{
          id: string
          createdAt: string
          updatedAt: string
          clientId: string | null
          connector_name: string
          connectionId: string | null
          raw: any
          unified: any
        }>(
          sql`SELECT * FROM ${sql.identifier([
            entityName,
          ])} WHERE "clientId" = ${endUser?.id ?? null}`,
        )
        yield res.rows.map((row) =>
          postgresHelpers._op('data', {
            data: {
              entityName,
              entity: row.unified,
              raw: row.raw,
              id: row.id,
              connectorName: 'postgres',
              connection_id: row.connectionId ?? undefined,
            },
          }),
        )
      }

      const rawPool = await rawClient.getPool()
      for (const [_entityName, _query] of Object.entries(sourceQueries ?? {})) {
        const entityName = _entityName as keyof NonNullable<
          typeof sourceQueries
        >
        const queryState = state[_entityName as keyof typeof state]
        if (!_query) {
          return
        }
        // If schema is known, we can use prepared statements instead. But in this case
        // we do not know the schema
        const query = handlebars.compile(_query)({
          ...queryState,
          endUserId: endUser?.id,
        })

        const res = await rawPool.query<{id?: string; modifiedAt?: string}>(
          rawClient.sql([query] as unknown as TemplateStringsArray),
        )
        const lastRow = res.rows[res.rows.length - 1]

        yield R.compact([
          ...res.rows.map((row) =>
            postgresHelpers._op('data', {
              data: {
                entityName,
                id: `${row.id}`,
                entity: row,
                connectorName: 'postgres', // is this right?
              },
            }),
          ),
          lastRow?.modifiedAt &&
            lastRow.id &&
            (postgresHelpers._opState({
              invoice: {
                lastModifiedAt: lastRow.modifiedAt,
                lastRowId: lastRow.id,
              },
            }) as never), // Temp hack...
        ])
      }
    }

    return rxjs
      .from(iterateEntities())
      .pipe(
        Rx.mergeMap((ops) =>
          rxjs.from([...ops, postgresHelpers._op('commit')]),
        ),
      )
  },
  destinationSync: ({endUser, source, settings: {databaseUrl}}) => {
    console.log('[destinationSync] Will makePostgresClient', {
      // databaseUrl,
      // migrationsPath: __dirname + '/migrations',
      endUser,
    })
    // TODO: Probably need to require these sql files to work... turn them into js files

    const {getPool} = makePostgresClient({
      databaseUrl,
      migrationsPath: __dirname + '/migrations',
      migrationTableName: '_migrations',
    })
    let batches: Record<string, Array<{id: string; [k: string]: unknown}>> = {}

    const migrationRan: Record<string, boolean> = {}
    async function runMigration(pool: DatabasePool, tableName: string) {
      console.log('will run migration for', tableName)
      if (migrationRan[tableName]) {
        return
      }
      migrationRan[tableName] = true
      // Where do we want to put data? Not always public...
      await setupTable({pool, tableName})
    }

    let agConnectionCreatedForResource = false;

    return handlersLink({
      data: async (op) => {
        const {
          data: {id, entityName, ...data},
        } = op
        const tableName = entityName
        const batch = batches[tableName] ?? []
        batches[tableName] = batch

        const rowToInsert: Record<string, unknown> = {
          // This is really not ideal. Maybe this should be a resource level seteting
          // about how we want to "normalize"?
          // Or be provided by the Operation itself?
          ...(typeof data.entity === 'object' &&
          data.entity &&
          'raw' in data.entity
            ? {...data.entity}
            : {raw: data.entity}),
          id,
          clientId: endUser?.id ?? null,
          // connectionId: source?.id,
          connectionId: source?.id + '',
          isOpenInt: true,
        }

        const isAgMode = 
          endUser?.orgId === 'org_2nJZrA4Dk8i3wszhm6PsP3M2Vwy' ||
          endUser?.orgId === 'org_2lcCCimyICKI8cpPNQt195h5zrP' ||
          endUser?.orgId === 'org_2ms9FdeczlbrDIHJLcwGdpv3dTx'

        // TODO: Remove when we have support for links custom upserts
        if(isAgMode) {
          console.log('Inserting record for AG');
          if (tableName === 'IntegrationAtsJob') {
            rowToInsert['external_job_id'] = data.entity?.raw?.id || '';
          } else if (tableName === 'IntegrationAtsCandidate') {
            rowToInsert['opening_external_id'] = data.entity?.raw?.id || '';
            rowToInsert['candidate_name'] = data.entity?.raw?.first_name + ' ' + data.entity?.raw?.last_name || '';
          } else if (tableName === 'IntegrationAtsJobOpening') {
            rowToInsert['opening_external_id'] = data.entity?.raw?.opening_id || '';
            rowToInsert['job_id'] = data.entity?.raw?.job_id || '';
          } else if (tableName === 'IntegrationAtsOffer') {
            // Note: These fields seemed duplicated from the nested objects
            rowToInsert['opening_external_id'] = data.entity?.raw?.opening?.id || '';
            // field does not exist in the offer object
            rowToInsert['candidate_name'] = ''
          }

          if(!agConnectionCreatedForResource) {
            console.log('will create ag connection record for', entityName);
            const integrationConnectionRecord = {
              id: source?.id + '',
              createdAt: new Date().toISOString(),
              updatedAt: new Date().toISOString(),
              clientId: endUser?.id,
              source: 'OpenInt',
              profile: 'Ats',
              provider: source?.connectorName, // lever, greenhouse etc
              label: source?.connectorName, // lever, greenhouse etc
            }
            const query = await upsertByIdQuery('IntegrationConnection', [integrationConnectionRecord], {
              primaryKey: ['id'],
            })

            if(!query) {
              console.log('[destinationSync] Could not AG create integration connection record');
              return;
            }

            // replace all instances of "Ats" with "ATS"
            const agMappings = [
              {from: 'client_id', to: 'clientId'},
              {from: 'created_at', to: 'createdAt'},
              {from: 'updated_at', to: 'updatedAt'},
              {from: 'is_open_int', to: 'isOpenInt'},
              {from: 'connection_id', to: 'connectionId'},
              ...agTableMappings,
            ]
            let sqlQuery = query.sql
            // Use a for loop to replace all camelCase table and column names with snake_case
            for (const mapping of agMappings) {
              const regex = new RegExp(`"${mapping.from}"`, 'g')
              sqlQuery = sqlQuery.replace(regex, `"${mapping.to}"`)
            }

            const pool = await getPool();
            await pool.query({
              sql: sqlQuery,
              values: query.values,
              type: query.type,
            });
            agConnectionCreatedForResource = true;
          } 
        }
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        batch.push(rowToInsert as any)
        return rxjs.of(op)
      },
      commit: async (op) => {
        const size = R.values(batches)
          .map((b) => b.length)
          .reduce((a, b) => a + b, 0)
        if (size === 0) {
          return op
        }
        const pool = await getPool()
        await Promise.all(
          Object.keys(batches).map((eName) => runMigration(pool, eName)),
        )

        console.log(`[postgres] Will commit ${size} entities`)
        await pool.transaction((client) =>
          Promise.all(
            R.pipe(
              batches,
              R.toPairs,
              R.map(([eName, batch]) =>
                upsertByIdQuery(eName, batch, {
                  primaryKey: [
                    'id',
                    // TODO: remove when we introduce dynamic column names
                    // 'connectionId'
                  ],
                }),
              ),
              R.compact,
              R.map((query) => {
                // TODO: remove when we introduce dynamic column names
                // Replace all instances inconsistent table and column names before execution
                const agMappings = [
                  {from: 'client_id', to: 'clientId'},
                  {from: 'created_at', to: 'createdAt'},
                  {from: 'updated_at', to: 'updatedAt'},
                  {from: 'is_open_int', to: 'isOpenInt'},
                  {from: 'connection_id', to: 'connectionId'},
                  ...agTableMappings,
                ]

                let sqlQuery = query.sql
                // Use a for loop to replace all camelCase table and column names with snake_case
                for (const mapping of agMappings) {
                  const regex = new RegExp(`"${mapping.from}"`, 'g')
                  sqlQuery = sqlQuery.replace(regex, `"${mapping.to}"`)
                }

                // replace all instances of "Ats" with "ATS"
                sqlQuery = sqlQuery.replace(/Ats/g, 'ATS')

                // console.log('sqlQuery', sqlQuery);
                return client.query({
                  sql: sqlQuery,
                  values: query.values,
                  type: query.type,
                })
              }),
            ),
          ),
        )
        batches = {}
        console.log(`[postgres] Did commit ${size} entities`)
        return op
      },
    })
  },
} satisfies ConnectorServer<typeof postgresSchemas>

export default postgresServer
