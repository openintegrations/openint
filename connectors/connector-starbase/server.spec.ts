import {getTableName, sql} from 'drizzle-orm'
import {drizzle} from 'drizzle-orm/sqlite-proxy'
import {starbase} from 'drizzle-starbase'
import * as R from 'remeda'
import type {CustomerId} from '@openint/cdk'
import {rxjs, toCompletion} from '@openint/util'
import {inferTableFromMessage} from '../connector-postgres'
import type {RecordMessageBody} from './def'
import starbaseServer from './server'

const STARBASE_ENDPOINT = process.env['STARBASE_ENDPOINT']!
const STARBASE_AUTH_TOKEN = process.env['STARBASE_AUTH_TOKEN']!

/**
 * Not working due to nextjs issue
 */
const db = drizzle(...starbase(STARBASE_ENDPOINT, STARBASE_AUTH_TOKEN))

const destLink = starbaseServer.destinationSync({
  config: {},
  customer: {id: 'esur_12' as CustomerId, orgId: 'org_123'},
  settings: {
    databaseUrl: STARBASE_ENDPOINT,
  },
  source: {id: 'conn_sfdc_9287', connectorName: 'sfdc'},
  state: {},
})

// Helper to clear tables between tests
async function clearTable(tableName: string) {
  await db.run(sql`DELETE FROM ${sql.identifier(tableName)}`)
}

describe('standard schema', () => {
  test('destinationSync', async () => {
    const src = rxjs.from([
      {
        data: {
          entityName: 'vendor',
          id: 'vend_1123',
          entity: {
            raw: {AccountName: 'Cash'},
            unified: {name: 'Assets:Cash'},
          },
        },
        type: 'data' as const,
      },
      {type: 'commit' as const},
    ])
    await toCompletion(destLink(src))

    const res = await db.run(`SELECT * FROM vendor WHERE id = 'vend_1123'`)
    expect(res.rows?.[0]).toMatchObject({
      source_id: 'conn_sfdc_9287',
      id: 'vend_1123',
      customer_id: 'esur_12',
      connector_name: 'sfdc',
      unified: JSON.stringify({name: 'Assets:Cash'}),
      raw: JSON.stringify({AccountName: 'Cash'}),
    })
  })
})

const messages = [
  ['insert', [{stream: 'house', data: {id: 112, name: 'White'}}]],
  // This is currently failing due to the removal of `IF NOT EXISTS` clause in drizzle-kit
  // Need to figure out if we should add it back, probably so?
  [
    'insert multiple',
    [
      {stream: 'account', data: {id: 555, name: 'Bank'}},
      {stream: 'trxn', data: {id: 555}},
    ],
  ],
  [
    'upsert single key',
    [
      {
        stream: 'tre',
        data: {id: 5, name: 'B'},
        upsert: {key_columns: ['id']},
      },
    ],
  ],
  [
    'upsert composite keys',
    [
      {
        stream: 'transaction',
        data: {
          id: '1',
          intId: 2,
          amount: 1.23,
          is_deleted: false,
          raw: {key: 'value'},
          array_value: [{key: 'value'}],
          null_value: null,
          undefined_value: undefined, // undefined gets turned into null... but should we allowe for it?
          // created_at: '2021-01-01T00:00:00Z', // this colun also causes problem for now
          // uupdated_at: new Date(), // Should we allow non-json values?
        },
        upsert: {
          key_columns: ['id', 'is_deleted'],
        },
      },
    ],
  ],
  // [
  //   'upsert multiple single key',
  //   [
  //     {entityName: 'tre', entity: {id: 5, name: 'B'}, upsert: {key_columns: ['id']}},
  //     // Needs to be two batches..n in order to work due to postgres upsert cannot operate on the same row twice in one txn
  //     {entityName: 'tre', entity: {id: 5, name: 'B'}, upsert: {key_columns: ['id']}},
  //   ],
  // ],
  [
    'mixed',
    [
      {
        stream: 'good',
        data: {id: 5, name: 'B'},
        upsert: {key_columns: ['id']},
      },
      {
        stream: 'life',
        data: {id: 1, code: 'C'},
        upsert: {key_columns: ['id']},
      },
    ],
  ],
] satisfies Array<[string, Array<Omit<RecordMessageBody, 'id'>>]>

describe.each(messages)('custom schema: %s', (_, messages) => {
  const tables = messages.map((m) =>
    inferTableFromMessage(m as unknown as RecordMessageBody),
  )

  test('destinationSync', async () => {
    const src = rxjs.from([
      ...messages.map((m) => ({
        data: {id: '', ...m},
        type: 'data' as const,
      })),
      {type: 'commit' as const},
    ])
    await toCompletion(destLink(src))

    const uniqueTables = R.uniqBy(tables, (t) => getTableName(t))
    const tableRows = await Promise.all(
      uniqueTables.map((t) =>
        db.run(`SELECT * FROM ${getTableName(t)}`).then((r) => r.rows ?? []),
      ),
    )

    const rowsByTable = Object.fromEntries(
      uniqueTables.map((t, i) => [getTableName(t), tableRows[i]]),
    )

    expect(rowsByTable).toMatchSnapshot()
  })
})
