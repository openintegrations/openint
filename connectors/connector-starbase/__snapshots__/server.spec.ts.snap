// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom schema: insert destinationSync 1`] = `
{
  "house": [
    {
      "id": 112,
      "name": "White",
    },
  ],
}
`;

exports[`custom schema: insert infer table schema 1`] = `
[
  "CREATE TABLE "house" (
	"id" integer,
	"name" text
);
",
]
`;

exports[`custom schema: insert multiple destinationSync 1`] = `
{
  "account": [
    {
      "id": 555,
      "name": "Bank",
    },
  ],
  "trxn": [
    {
      "id": 555,
    },
  ],
}
`;

exports[`custom schema: insert multiple infer table schema 1`] = `
[
  "CREATE TABLE "account" (
	"id" integer,
	"name" text
);
",
  "CREATE TABLE "trxn" (
	"id" integer
);
",
]
`;

exports[`custom schema: mixed destinationSync 1`] = `
{
  "good": [
    {
      "id": 5,
      "name": "B",
    },
  ],
  "life": [
    {
      "code": "C",
      "id": 1,
    },
  ],
}
`;

exports[`custom schema: mixed infer table schema 1`] = `
[
  "CREATE TABLE "good" (
	"id" integer,
	"name" text
);
",
  "CREATE UNIQUE INDEX "good_upsert_keys" ON "good" USING btree ("id");",
  "CREATE TABLE "life" (
	"id" integer,
	"code" text
);
",
  "CREATE UNIQUE INDEX "life_upsert_keys" ON "life" USING btree ("id");",
]
`;

exports[`custom schema: upsert composite keys destinationSync 1`] = `
{
  "transaction": [
    {
      "amount": 1.23,
      "array_value": [
        {
          "key": "value",
        },
      ],
      "id": "1",
      "intId": 2,
      "is_deleted": false,
      "null_value": null,
      "raw": {
        "key": "value",
      },
      "undefined_value": null,
    },
  ],
}
`;

exports[`custom schema: upsert composite keys infer table schema 1`] = `
[
  "CREATE TABLE "transaction" (
	"id" text,
	"intId" integer,
	"amount" double precision,
	"is_deleted" boolean,
	"raw" jsonb,
	"array_value" jsonb,
	"null_value" jsonb,
	"undefined_value" jsonb
);
",
  "CREATE UNIQUE INDEX "transaction_upsert_keys" ON "transaction" USING btree ("id","is_deleted");",
]
`;

exports[`custom schema: upsert single key destinationSync 1`] = `
{
  "tre": [
    {
      "id": 5,
      "name": "B",
    },
  ],
}
`;

exports[`custom schema: upsert single key infer table schema 1`] = `
[
  "CREATE TABLE "tre" (
	"id" integer,
	"name" text
);
",
  "CREATE UNIQUE INDEX "tre_upsert_keys" ON "tre" USING btree ("id");",
]
`;
